modules/service-autoscaling/main.tf

```diff
@@ -0,0 +1,84 @@
+terraform {
+  required_version = ">= 0.11.11"
+}
+
+resource "aws_appautoscaling_target" "service_appautoscaling_target" {
+  max_capacity       = "${var.max_capacity}"
+  min_capacity       = "${var.min_capacity}"
+  resource_id        = "service/${var.cluster_name}/${var.service_name}"
+  scalable_dimension = "ecs:service:DesiredCount"
+  service_namespace  = "ecs"
+}
+
+resource "aws_appautoscaling_policy" "queue_backlog_scaling_policy" {
+  name               = "${var.service_name}-${var.queue_name}-queue-backlog-scaling-policy"
+  policy_type        = "TargetTrackingScaling"
+  resource_id        = "${aws_appautoscaling_target.service_appautoscaling_target.resource_id}"
+  scalable_dimension = "${aws_appautoscaling_target.service_appautoscaling_target.scalable_dimension}"
+  service_namespace  = "${aws_appautoscaling_target.service_appautoscaling_target.service_namespace}"
+
+  target_tracking_scaling_policy_configuration {
+    target_value       = "${var.target_value}"
+    scale_in_cooldown  = "${var.scalein_cooldown}"
+    scale_out_cooldown = "${var.scaleout_cooldown}"
+
+    customized_metric_specification {
+      metric_name = "QueueBacklog"
+      namespace   = "AWS/ECS"
+      statistic   = "${var.target_value_statistic}"
+
+      dimensions {
+        name  = "ClusterName"
+        value = "${var.cluster_name}"
+      }
+
+      dimensions {
+        name  = "ServiceName"
+        value = "${var.service_name}"
+      }
+
+      dimensions {
+        name  = "QueueName"
+        value = "${var.queue_name}"
+      }
+    }
+  }
+}
+
+resource "aws_appautoscaling_policy" "queue_requires_consumer_scaling_policy" {
+  name               = "${var.service_name}-${var.queue_name}-queue-requires-consumer-scaling-policy"
+  policy_type        = "StepScaling"
```

Comment:
Is there a reason to make this a variable as well with a default to StepScaling?

Reply:
No, since only StepScaling and TargetTrackingScaling policies are valid ECS autoscaling policies. I can update this if that ever changes, though.